src/libre/re_char_class.c

```diff
@@ -0,0 +1,802 @@
+{
+	char_filter *f = NULL;
+	assert(r != NULL);
+
+	switch (r->t) {
+	case AST_RANGE_ENDPOINT_LITERAL:
+		if (from != NULL) { *from = r->u.literal.c; }
+		if (to != NULL) { *to = r->u.literal.c; }
+		break;
+	case AST_RANGE_ENDPOINT_CHAR_TYPE:
+	{
+		unsigned negated = 0;
+		char_filter_for_char_type_id(r->u.char_type.id, &f, &negated);
+		char_endpoints_for_filter(f, negated, from, to);
+		break;
+	}
+	case AST_RANGE_ENDPOINT_CHAR_CLASS:
+		f = char_filter_for_class_id(r->u.char_class.id);
+		assert(f != NULL);
+		char_endpoints_for_filter(f, 0, from, to);
+		break;
+	default:
+		assert(0);
+	}
+}
+
+static void
+char_endpoints_for_filter(char_filter *f, unsigned negated,
+    unsigned char *low, unsigned char *high)
+{
+	unsigned i, highest;
+	int has_l = 0, has_h = 0;
+
+	assert(f != NULL);
+	for (i = 0; i < 256; i++) {
+		int match = f(i);
+		if ((negated && match) || (!negated && !match)) {
+			continue;
+		}
+		if (low != NULL && !has_l) {
+			has_l = 1;
+			*low = i;
+			if (high == NULL) { return; }
+		}
+
+		has_h = 1;
+		highest = i;
+	}
+
+	if (high != NULL && has_h) {
+		*high = highest;
+	}
+}
+
+static void
+cc_add_range(struct re_char_class *cc, 
+    const struct ast_range_endpoint *from,
+    const struct ast_range_endpoint *to)
+{
+	unsigned char lower, upper;
+	unsigned char i;
+
+	re_char_class_endpoint_span(from, &lower, NULL);
+	re_char_class_endpoint_span(to, NULL, &upper);
+
+	assert(cc != NULL);
+	assert(lower <= upper);
+	for (i = lower; i <= upper; i++) {
+		cc_add_byte(cc, i);		
+	}
+}
+
+static char_filter *
+char_filter_for_class_id(enum ast_char_class_id id)
+{
+	char_filter *filter = NULL;
+	switch (id) {
+	case AST_CHAR_CLASS_ALNUM:
+		filter = isalnum; break;
+	case AST_CHAR_CLASS_ALPHA:
+		filter = isalpha; break;
+	case AST_CHAR_CLASS_ANY:
+		break;
+	case AST_CHAR_CLASS_ASCII:
+		/* filter = isascii; */
+		break;
+	case AST_CHAR_CLASS_BLANK:
+		/* filter = isblank; */
+		break;
+	case AST_CHAR_CLASS_CNTRL:
+		filter = iscntrl; break;
+	case AST_CHAR_CLASS_DIGIT:
+		filter = isdigit; break;
+	case AST_CHAR_CLASS_GRAPH:
+		filter = isgraph; break;
+	case AST_CHAR_CLASS_LOWER:
+		filter = islower; break;
+	case AST_CHAR_CLASS_PRINT:
+		filter = isprint; break;
+	case AST_CHAR_CLASS_PUNCT:
+		filter = ispunct; break;
+	case AST_CHAR_CLASS_SPACE:
+		filter = isspace; break;
+	case AST_CHAR_CLASS_SPCHR:
+		/* filter = isspchr; */
+		break;
+	case AST_CHAR_CLASS_UPPER:
+		filter = isupper; break;
+	case AST_CHAR_CLASS_WORD:
+		/* filter = isword; */
+		break;
+	case AST_CHAR_CLASS_XDIGIT:
+		filter = isxdigit; break;
+	default:
+		fprintf(stderr, "(MATCH FAIL)\n");
+		assert(0);
+	}
+	return filter;
+}
+
+static int
+cc_add_named_class(struct re_char_class *cc, enum ast_char_class_id id)
+{
+	unsigned i;
+	char_filter *filter = NULL;
+	assert(cc != NULL);
+
+	filter = char_filter_for_class_id(id);
+
+	if (filter == NULL) {
+		return 0;
+	}
+
+	for (i = 0; i < 256; i++) {
+		if (filter(i)) { cc_add_byte(cc, i); }
```

Comment:
I think iterating 0..UCHAR_MAX here is the wrong approach; the named classes are provided as DFA currently because their contents need to handle both UTF-8 sequences and ASCII.

Perhaps the AST ought to think in terms of codepoints, rather than bytes. I'm not sure.

My instinct is that the "filter" functions should not exist, because we shouldn't be running things in order to construct them. However we express them, I'd hope these sets can be constructed ahead of time.